### PR TITLE
v0.1.9

### PR DESCRIPTION
**Pull request by BrikerMan**
add `AVCNNModel`, `KMaxCNNModel`, `RCNNModel`, `AVRNNModel`, `DropoutBGRUModel`, `DropoutAVRNNModel` model to classification task.

### CODE REVIEW

**Comment by alexwwang** [comment]
I reproduced the dim mismatch error during saved model loading. 
I double checked the `KMaxPooling` layer in the referenced code file and compared it with the original one defined in VDCNN model. There are two tiny differences between these two editions and I suspect it's the reason arousing this shape mismatch error. 
I would try to modify the layer function and  this model tomorrow to prove it or not.

**Comment by alexwwang** [comment]
After one day investigation, I tend to suspect this would be caused by an unknown bug in Keras during loading a model with custom layer. As long as I alternate the `KMaxPooling` layer with factory `GlobalMaxPooling1D` layer, the error disappears. 
But more works need to be done to find out the reason and solution.

At least two directions could be explored:
- Complete the KMaxPooling layer define with more detailed class method. ( I've implemented `compute_out_shape`, `build` and `get_config`, but the error stays.)
- Try to save and load the model in two steps, structure first and weights second. However I am not sure if this would work.

Some references: https://github.com/keras-team/keras/issues/8612

**Comment by BrikerMan** [comment]
I have tried the `model.model.load_weights('model/model.model')`, it works just fine. But we have to refactor the save and load function.

**Comment by alexwwang** [comment]
> I have tried the `model.model.load_weights('model/model.model')`, it works just fine. But we have to refactor the save and load function.

You tried it? It works?
I think it's a good news!
Then I would make a new PR first to update my modification to the layer and model implementation.
And then let's refactor the save/load process.

**Comment by BrikerMan** [comment]
I have tried this and this works fine, I think this means we could build the model first then load the saved weight later.

```python
model = KMaxCNNModel()
model.fit(x, y, epochs=2)
model.save('model')

# works fine
model.model.load_weights('model/model.model')

# Error
KMaxCNNModel.load_model('model')
```

**Comment by alexwwang** [comment]
> I have tried this and this works fine, I think this means we could build the model first then load the saved weight later.
> 
> ```python
> model = KMaxCNNModel()
> model.fit(x, y, epochs=2)
> model.save('model')
> 
> # works fine
> model.model.load_weights('model/model.model')
> 
> # Error
> KMaxCNNModel.load_model('model')
> ```

Great!
I've refactored the s/l part and am testing the model again.
After its done, I'll make another PR.

**Comment by alexwwang** [comment]
> I have tried this and this works fine, I think this means we could build the model first then load the saved weight later.
> 
> ```python
> model = KMaxCNNModel()
> model.fit(x, y, epochs=2)
> model.save('model')
> 
> # works fine
> model.model.load_weights('model/model.model')
> 
> # Error
> KMaxCNNModel.load_model('model')
> ```

I still got an error when load the model and it said
`Expecting value: line 1 column 1 (char 0)`
Not sure what's wrong with it.

**Comment by alexwwang** [comment]
> I have tried this and this works fine, I think this means we could build the model first then load the saved weight later.
> 
> ```python
> model = KMaxCNNModel()
> model.fit(x, y, epochs=2)
> model.save('model')
> 
> # works fine
> model.model.load_weights('model/model.model')
> 
> # Error
> KMaxCNNModel.load_model('model')
> ```

Your test may be not the exact situation.
I am afraid you should test a new model from scratch, say create it from the structure saved and then load weights to it. 
This is what `load_model` is supposed to do but not load weights to an trained model.

**Comment by BrikerMan** [comment]
@alexwwang You are right, let me try a new model then load the weights to the new one.

**Comment by alexwwang** [comment]
> @alexwwang You are right, let me try a new model then load the weights to the new one.

I tried, the problem is not gone. :(

**Comment by BrikerMan** [comment]
Ok, let's release this version without the `KMaxCNNModel` model~

**Comment by alexwwang** [comment]
> Ok, let's release this version without the `KMaxCNNModel` model~

Well, but it's really annoying.
Do you have any idea on supporting customized layer definition and registration to this scheme? Just like what you've done on customized embedding layers.
Otherwise each time someone creates a new model with certain customized layers, you have to refactor the model, splitting the layers define section out.

**Comment by alexwwang** [comment]
I've got an idea.
Since you have tested to load weights to a compiled model, I think it would worth a try to change the load process,  rather than loading structure from json, but compile the model first and then load the weights.

**Comment by BrikerMan** [comment]
@alexwwang yes, maybe this will work.

**Comment by alexwwang** [comment]
I've solved it!
Make a pr now.
